src/NuGet.Clients/VsConsole/PowerShellHost/PowerShellHost.cs

```diff
@@ -372,95 +375,174 @@ private void UpdateWorkingDirectory()
+            // Process each framework/id/version once to avoid duplicate work
+            // Packages may have different dependendcy orders depending on the framework, but there is 
+            // no way to fully solve this across an entire solution so we make a best effort here.
+            foreach (var framework in packagesConfigInstalled.Keys.OrderByDescending(fw => fw, new NuGetFrameworkSorter()))
+            {
+                foreach (var package in packagesConfigInstalled[framework])
                 {
-                    // Packages may occur under multiple projects, but we only need to run it once.
-                    if (!finishedPackages.Contains(package))
+                    if (resolvedPackages.Add(package))
                     {
-                        finishedPackages.Add(package);
-
-                        try
+                        var dependencyInfo = await dependencyInfoResource.ResolvePackage(
+                            package,
+                            framework,
+                            NullLogger.Instance,
+                            CancellationToken.None);
+
+                        // This will be null for unrestored packages
+                        if (dependencyInfo != null)
                         {
-                            string pathToPackage = null;
+                            packagesToOrder.Add(new ResolverPackage(dependencyInfo, listed: true, absent: false));
+                        }
+                    }
+                }
+            }
 
-                            // If the package exists in both the global and packages folder, use the packages folder copy.
-                            if (packagesFolderPackages.Contains(package))
-                            {
-                                // Local package in the packages folder
-                                pathToPackage = packagePathResolver.GetInstalledPath(package);
-                            }
-                            else
-                            {
-                                // Global package
-                                pathToPackage = fallbackResolver.GetPackageDirectory(package.Id, package.Version);
-                            }
+            // Prioritize packages.config folder paths over global folder paths for legacy.
+
+            // Order packages by dependency order
+            var sortedPackages = ResolverUtility.TopologicalSort(packagesToOrder);
+            foreach (var package in sortedPackages)
+            {
+                if (finishedPackages.Add(package))
+                {
+                    // Find the package path in the packages folder
+                    var pathToPackage = packagePathResolver.GetInstalledPath(package);
 
-                            if (!string.IsNullOrEmpty(pathToPackage))
+                    ExecuteInitPs1(pathToPackage, package);
+                }
+            }
+        }
+
+        private void ExecuteInitPs1ForBuildIntegrated(
```

Comment:
There are no async calls in this method, there would be a compiler warning if it was async.